File: project-euler.java/src/main/java/euler/solutions/_26.java
package euler.solutions;

import euler.common.BaseSolution;
import gnu.trove.list.TIntList;
import gnu.trove.list.array.TIntArrayList;

public class _26 extends BaseSolution {

    @Override
    public long solve() {
        int maxD = 0, maxLen = 0;

        for (int d = 1; d <= 1000; d++) {
            int length = unitFractionCycleLength(d);
            if (length > maxLen) {
                maxLen = length;
                maxD = d;
            }
        }

        return maxD;
    }

    private int unitFractionCycleLength(int d) {
        int result = 1;
        TIntList prev = new TIntArrayList();

        while (result != 0)
        {
            result = result * 10 % d;
            int idx = prev.indexOf(result);

            if (idx != -1) {
                return (prev.size() - idx);
            }

            prev.add(result);
        }

        return 0;
    }
}

File: project-euler.go/solutions/_26/26.go
package _26

import "project-euler/common/numbers"

func Solve() int64 {
	maxD, maxLen := 0, 0

	for d := 1; d <= 1000; d++ {
		length := unitFractionCycleLength(d)
		if length > maxLen {
			maxLen = length
			maxD   = d
		}
	}

	return int64(maxD)
}

func unitFractionCycleLength(d int) int {
	result := 1
	prev := []int{}

	for result != 0 {
		result = result * 10 % d
		idx := numbers.IndexOf(prev, result)

		if idx != -1 {
			return len(prev) - idx
		}

		prev = append(prev, result)
	}

	return 0
}
